Improve quality checks and lint the code

pawamoy opened the pull request:
Issue: #38

pawamoy commented:
@ziima 

The available tox environments are now:

- `test`:
  Run all the Python/Django test environments.
- `check`:
  Run all the check environments.
- `parallel-test`:
  Run all the Python/Django test environments in parallel.
- `parallel-check`:
  Run all the check environments in parallel.
- `parallel-all`:
  Run all the environments in parallel.
- `build-docs`:
  Build the documentation locally.
- `check-setup`:
  Check that the package will be correctly installed and correctly rendered on PyPI.
- `check-bandit`:
  Run bandit tool on the code.
- `check-safety`:
  Check that the requirements versions do not have security vulnerabilities.
- `check-black`:
  Run black tool on the code.
- `check-isort`:
  Check the imports order.
- `check-pylint`:
  Check the code style.
- `check-docs-spell`:
  Check the spelling in the documentation.
- `check-docs-link`:
  Check that the links written in documentation are valid.
- `run-black`:
  Run black tool on the code.
- `run-isort`:
  Check the imports order.
- `run-bumpversion`:
  Check the imports order.
- `report`:
  Create coverage report.
- `clean`:
  Delete coverage report.
- `codacy`:
  Upload coverage report to codacy.

Also:

- Configuration for tools has been moved in respective files in `config` directory. It was before all in `setup.cfg`.
- Prospector has been changed for pylint. Pylint job is allowed to fail on CI.
- MANIFEST.in has been simplified, and will reduce the size of the built packages.
- Line length set to 120.
- Pandas and Django sections removed from isort config.
- Ran black and isort on the code.
- Travis config updated to reflect the changes.

ziima commented on config/.isort.cfg:
```diff
@@ -0,0 +1,11 @@
+default_section = THIRDPARTY
+forced_separate = test_appsettings
+known_first_party = appsettings
+sections = FUTURE,STDLIB,THIRDPARTY,FIRSTPARTY,LOCALFOLDER
```
* I believe the `sections` are set to default, so we can delete them.
 * `forced_separate = test_appsettings` seems to be unused.

ziima commented on tox.ini:
```diff
@@ -123,41 +152,62 @@ deps =
+deps =
+	bumpversion
+commands =
+	bumpversion --config-file config/bumpversion.conf {posargs}
```
`config/bumpversion.conf` does not exist.

pawamoy replied:
Oops :fearful:

pawamoy commented:
> I'm not sure that moving the config files into config directory is a good move. For example various editor plugins will not find their configuration in there. I don't really mind using setup.cfg for these.

I think you generally can configure these plugins to give the location of the config file. I myself use PyCharm and you can configure your external tools and set up parameters for execution so everything is possible. However, I don't use any external tool in my IDE, so I reverted that back, just in case. Config is now in `setup.cfg`, except for black which has its config in `pyproject.toml`.

> I'm not very favor of pylint for the number of false positives it produces. I can recommend flake8 instead.

I agree. Replaced pylint with flake8.

> I have never thought about using tox for anything else but code checks and tests. What's your experience with using tox for code modification?

Actually a lot of people seem to use tox as some kind of `make` tool. So not only I'm used to write test environments and "check" environments, but also environments that will perform some action on the code. It usually works well and can reduce the friction for new contributors, as everything you need is in tox and tox takes care of installing the tools needed.

pawamoy commented:
`isort` and `black` now conflict with each other on the trailing comma in imports... :confused:

ziima commented on MANIFEST.in:
```diff
@@ -1,22 +1,20 @@
+prune scripts
+prune tests
 
+exclude * .*
```
This seems to be too intrusive? Which files do you intend to exclude? I would definitely keep `docs`, `config` and `requirements`. `tests` should be excluded by the `setup.py` itself.

pawamoy replied:
Honestly, I intend to exclude as much as I can. 

- `.cookiecutterrc`:  development related, not needed
- `.coverage`: test-execution related, not needed
- `.editorconfig`: development related, not needed
- `pyproject.toml`: only used for black right now, so test-related, not needed if we don't include tests
- `runtests.py`: test-related, not needed if we don't include tests
- `tox.ini`: dev/test-related, not needed if we don't include tests
- `.travis.yml`: ci-related, not needed

I searched and searched over the internet but I can't find any post about Python packaging emphasizing the need to include development-related files in sdists/wheels. And I personally don't see what are the benefits would be.

- `tests/`: I never ran tests against installed packages, and I never heard of anyone doing it. Also, why should they be excluded by `setup.py` and not by `MANIFEST.in`? I find it more consistent to put that rule in the manifest.
- `docs/`: I never read the docs of an installed packages (from the local, installed files), and I never heard of anyone doing it, but for this one I agree we could add it (the built version, not the raw RST files), because it still could be useful when in offline mode.
- `config/`: we don't have this directory anymore so I should remove that line from MANIFEST.in :slightly_smiling_face: 
- `requirements/`: only useful for development/CI purposes as it's used by tox only. All the "production" requirements are listed in `setup.py`.

Of course these are personal opinions, so please prove me wrong with examples :slightly_smiling_face:. Or maybe I'm missing something obvious :confused:

ziima replied:
By default, source distribution only includes files handled by `setup.py`, so `MANIFEST.in` should only list files you want to include. As a result, you don't need to exclude files you don't want.

The `MANIFEST.in` on master looks good enough to me. If you don't want some of the files, you should be fine with just removing those lines from the manifest. Tests doesn't seem to be included by `setup.py`, so they doesn't need to be mentioned in manifest neither, they will not be installed.

I'm not quite sure about details of source version, but in my opinion they should include all files needed to build the application and its docs (I don't use them either, but some admins complain about missing docs). There is somewhat trouble with tests as they should be run as part of the build process, e.g. building a Debian package, but I'm not quite sure, how it should affect the source distribution.

pawamoy replied:
I think I explicitly excluded files, even if it was not necessary, because the old `python setup.py check` command would complain about versioned files not included by the manifest. Maybe it's different now with `twine check`. Will see that.

ziima replied:
I doesn't seem to be a problem any more. Let me know if you want me to check another PR.

pawamoy commented on MANIFEST.in:
```diff
@@ -1,22 +1,20 @@
+
+include setup.py
+include setup.cfg
+include MANIFEST.in
```
But were excluded by the previous line, `exclude * .*`, so I had to explicitly include them.

ziima replied:
Then the `exclude` clearly doesn't do what it should.

ziima replied:
Well, it seems useless to exclude all the files and then include defaults.

pawamoy commented:
Closing in favor of #45.
